Flaky `test_background_process` on macOS, and drop unused coverage from the default test run

Two independent cleanups in the `core` unit-test job. They surfaced together while triaging a red macOS
build on #501, but **they are not cause and effect** — I measured, see "Coverage is not the cause" below.

## 1. `test_background_process` is timing-fragile and flakes on macOS

[`pimm/tests/test_world.py:216`](../blob/main/pimm/tests/test_world.py#L216) guesses at two durations
instead of waiting for the thing it cares about:

```python
world.start_in_subprocess(dummy_process)

time.sleep(0.2)          # hope the child has started
assert len(world.background_processes) == 1

world._stop_event.set()
world.background_processes[0].join(timeout=0.5)   # hope it has exited
assert not world.background_processes[0].is_alive()
assert world.background_processes[0].exitcode == 0
```

Observed failure — `core (macos-latest, 3.13)` on
[run 30258019123](https://github.com/Positronic-Robotics/positronic/actions/runs/30258019123/job/89951072234),
`1 failed, 818 passed, 7 skipped`:

```
>  assert not world.background_processes[0].is_alive()
E  AssertionError: assert not True
E   +  where True = <SpawnProcess name='dummy_process' pid=2487 started daemon>.is_alive
```

### Mechanism

macOS uses the `spawn` start method, so the child re-imports the world from scratch (pytest, `pimm`, the test
module) rather than inheriting a fork. Once it *is* running, exit is quick and not the problem: `dummy_process`
polls every 10 ms, and `_stop_event` is already set by the time a late child first reads it, so it drops out of
the loop on its first iteration.

The whole risk sits in startup. The test gives the child 0.2 s + 0.5 s = **0.7 s to spawn, import, and exit**.
On a loaded GitHub macOS runner that budget is simply too tight — the failure is a child that hadn't finished
importing yet, not a child that failed to stop.

Linux forks, which is far cheaper, which is why the three Ubuntu jobs stay green.

### It is a rare flake, not a chronic one

Exactly one failure across the last 60 `unit-test.yaml` runs — this one. `macos-latest` has been in the matrix
since #408, so that is a real baseline rather than a young job. It also has nothing to do with the PR it landed
on: #501 touches no file under `pimm/`.

### Suggested fix

Replace both fixed waits with a poll against a deadline — wait for the child to actually be up before setting
the stop event, then poll `is_alive()` until a generous deadline instead of a single `join(0.5)`. Correctness
of the assertion is unchanged; it just stops encoding "0.7 s is enough to spawn a Python process".

## 2. Drop coverage from the default test run

Coverage runs on every invocation via `addopts` in `pyproject.toml`, and nothing consumes the result:

```toml
addopts = [
    "--cov=pimm",
    "--cov=positronic",
    "--cov-report=term-missing",
    ...
]
```

Evidence it is unused:

- No codecov integration, no threshold, no `--cov-fail-under` — nothing gates on the number.
- `unit-test.yaml` writes an HTML artifact and a job summary that nobody reads.
- Every other test job already opts out: `--no-cov` in the two `lerobot` jobs and in `libero-e2e.yaml`.
- `CLAUDE.md` instructs contributors to pass `--no-cov` for *all* local runs, so it isn't used locally either.

### Measured cost

Full suite on an M-series Mac, `uv run --locked --extra dev pytest`:

| | wall time |
|---|---|
| `--no-cov` | 66.6 s |
| with coverage | 72.6 s |

About **9%** on the suite, plus ~1.3 s of fixed `pytest-cov` session startup on every invocation (which
dominates when running a single test: 0.56 s → 1.9 s), plus the `coverage report` and artifact-upload steps in
CI.

### What to change

- [ ] Remove the three `--cov*` entries from `addopts` in `pyproject.toml`
- [ ] Drop the coverage report / HTML artifact steps from the `core` job in `unit-test.yaml`
- [ ] Update `CLAUDE.md` — the documented commands carry a `--no-cov` that is no longer needed
- [ ] Keep `coverage` / `pytest-cov` in the `dev` extra so it can still be run on demand locally

### Coverage is not the cause

Worth recording, because the two items look causally linked and aren't. Coverage instruments spawned children,
so it was a natural suspect for the flake. It isn't — the test's own `call` duration is the same either way:

```
--- with coverage ---            --- no coverage ---
0.49s test_background_process    0.46s test_background_process
0.46s test_background_process    0.47s test_background_process
0.46s test_background_process    0.47s test_background_process
```

(A single-test run *does* look ~3x slower with coverage, but that is `pytest-cov` session startup, not the
test.) Removing coverage will not fix the flake, and fixing the flake does not depend on removing coverage —
so they can land in either order.

--- a/.github/workflows/libero-e2e.yaml
+++ b/.github/workflows/libero-e2e.yaml
@@ -41,7 +41,7 @@ jobs:
       - name: Replay a LIBERO demo through the env server (software rendering)
         env:
           MUJOCO_GL: osmesa
-        run: uv run pytest positronic/simulator/libero/tests/test_e2e.py --no-cov -rs
+        run: uv run pytest positronic/simulator/libero/tests/test_e2e.py -rs
 
       # Reuses the LIBERO checkout the replay above cloned (``~/.cache/positronic/libero/src``). Validates the
       # command transform (FK/IK + each controller's set_goal inverse) and the gripper normalization against

--- a/.github/workflows/unit-test.yaml
+++ b/.github/workflows/unit-test.yaml
@@ -31,21 +31,10 @@ jobs:
       - name: Install dependencies
         run: uv sync --locked
 
-      - name: "Run tests (coverage: term, html)"
+      - name: Run tests
         env:
           PYTHONPATH: ${{ env.PYTHONPATH }}:$PWD
-        run: uv run pytest --cov-report=html
-
-      - name: Coverage summary (job summary)
-        if: always()
-        run: uv run coverage report -m >> $GITHUB_STEP_SUMMARY
-
-      - name: Upload coverage HTML artifact
-        if: always()
-        uses: actions/upload-artifact@v4
-        with:
-          name: coverage-html-core-${{ matrix.os }}-${{ matrix.python }}
-          path: htmlcov
+        run: uv run pytest
 
   packaging:
     runs-on: ubuntu-latest
@@ -81,7 +70,7 @@ jobs:
       - name: Run lerobot vendor tests
         env:
           PYTHONPATH: ${{ env.PYTHONPATH }}:$PWD
-        run: uv run pytest positronic/vendors/lerobot_0_3_3/tests --no-cov
+        run: uv run pytest positronic/vendors/lerobot_0_3_3/tests
 
   lerobot-latest:
     runs-on: ubuntu-latest
@@ -101,7 +90,7 @@ jobs:
       - name: Run lerobot vendor tests
         env:
           PYTHONPATH: ${{ env.PYTHONPATH }}:$PWD
-        run: uv run pytest positronic/vendors/lerobot/tests --no-cov
+        run: uv run pytest positronic/vendors/lerobot/tests
 
   lockfile-portability:
     strategy:

--- a/CLAUDE.md
+++ b/CLAUDE.md
@@ -15,8 +15,8 @@
 
 # Commands
 - Every Python execution goes through `uv run --locked` — bare `python`/`pytest` bypasses the locked venv
-- Run tests: `uv run --locked pytest --no-cov`
-- Run single test file: `uv run --locked pytest path/to/test_file.py --no-cov`
+- Run tests: `uv run --locked pytest`
+- Run single test file: `uv run --locked pytest path/to/test_file.py`
 - Lint: `uv run --locked ruff check --fix .`
 - Format: `uv run --locked ruff format .`
 - Run any Python: `uv run --locked python script.py`

--- a/README.md
+++ b/README.md
@@ -321,7 +321,7 @@ repoints it at the tool.
 Run tests and linters from the root directory:
 
 ```bash
-uv run --locked pytest --no-cov
+uv run --locked pytest
 uv run --locked ruff check .
 uv run --locked ruff format .
 ```

--- a/docs/CONTRIBUTING.md
+++ b/docs/CONTRIBUTING.md
@@ -151,7 +151,7 @@ uv run pytest
 With coverage report:
 
 ```bash
-uv run pytest --cov=positronic --cov=pimm --cov-report=term-missing
+uv run pytest --cov --cov-report=term-missing
 ```
 
 ## Pull Request Guidelines

--- a/pimm/tests/test_world.py
+++ b/pimm/tests/test_world.py
@@ -2,6 +2,7 @@ import logging
 import multiprocessing as mp
 import struct
 import time
+from functools import partial
 from queue import Empty, Full
 from unittest.mock import Mock, patch
 
@@ -228,18 +229,13 @@ class TestVirtualClock:
             assert isinstance(world.clock, SystemClock)
 
 
-# Module scope: `start_in_subprocess` pickles the loop to reach a spawned child, and a class defined
-# inside the test would not pickle.
-class HeartbeatLoop:
+# Module scope: `start_in_subprocess` pickles the loop to reach a spawned child, and a definition inside
+# the test would not pickle.
+def heartbeat_loop(emitter, stop_reader, clock):
     """Control loop announcing every iteration of its body."""
-
-    def __init__(self, emitter: SignalEmitter):
-        self.emitter = emitter
-
-    def run(self, stop_reader, clock):
-        while not stop_reader.read().data:
-            self.emitter.emit('beat')
-            yield Sleep(0.01)
+    while not stop_reader.read().data:
+        emitter.emit('beat')
+        yield Sleep(0.01)
 
 
 class TestWorld:
@@ -260,7 +256,7 @@ class TestWorld:
         with world:
             emitter, receiver = world.mp_pipes()
             assert isinstance(receiver, SignalReceiver)
-            world.start_in_subprocess(HeartbeatLoop(emitter).run)
+            world.start_in_subprocess(partial(heartbeat_loop, emitter))
 
             assert len(world.background_processes) == 1
 

--- a/pimm/tests/test_world.py
+++ b/pimm/tests/test_world.py
@@ -1,7 +1,6 @@
 import logging
 import multiprocessing as mp
 import struct
-import time
 from queue import Empty, Full
 from unittest.mock import Mock, patch
 
@@ -246,7 +245,6 @@ class TestWorld:
         with world:
             world.start_in_subprocess(dummy_process)
 
-            time.sleep(0.2)  # Some time to let the process run
             assert len(world.background_processes) == 1
 
             # We have to set the private event manually, because out of the scope of the context manager

--- a/pimm/tests/test_world.py
+++ b/pimm/tests/test_world.py
@@ -1,6 +1,7 @@
 import logging
 import multiprocessing as mp
 import struct
+import time
 from queue import Empty, Full
 from unittest.mock import Mock, patch
 
@@ -29,6 +30,18 @@ def dummy_process(stop_reader, clock):
     """A simple background process that runs until stopped."""
     while not stop_reader.read().data:
         yield Sleep(0.01)
+
+
+class HeartbeatProcess:
+    """Background process announcing every iteration of its control loop."""
+
+    def __init__(self, emitter: SignalEmitter):
+        self.emitter = emitter
+
+    def run(self, stop_reader, clock):
+        while not stop_reader.read().data:
+            self.emitter.emit('beat')
+            yield Sleep(0.01)
 
 
 class DummyControlSystem(ControlSystem):
@@ -243,16 +256,22 @@ class TestWorld:
         """Test that background processes will run simple control loop."""
         world = World()
         with world:
-            world.start_in_subprocess(dummy_process)
+            emitter, receiver = world.mp_pipes()
+            assert isinstance(receiver, SignalReceiver)
+            world.start_in_subprocess(HeartbeatProcess(emitter).run)
 
             assert len(world.background_processes) == 1
+
+            # Stopping before the child reaches its loop body would leave the loop untested: the generator
+            # sees a set event on its first condition and returns without ever running the body.
+            deadline = time.monotonic() + 30
+            while receiver.read() is None:
+                assert time.monotonic() < deadline, 'background process never entered its control loop'
+                time.sleep(0.01)
 
             # We have to set the private event manually, because out of the scope of the context manager
             # we can't access exit code of the process
             world._stop_event.set()
-            # The child is spawned, so it boots a fresh interpreter and imports this module before it can
-            # observe the stop event — on a slow runner that outlasts any tight deadline. `join` returns the
-            # moment it exits, so a generous ceiling costs a passing run nothing.
             world.background_processes[0].join(timeout=30)
             assert not world.background_processes[0].is_alive()
             assert world.background_processes[0].exitcode == 0

--- a/pimm/tests/test_world.py
+++ b/pimm/tests/test_world.py
@@ -32,18 +32,6 @@ def dummy_process(stop_reader, clock):
         yield Sleep(0.01)
 
 
-class HeartbeatProcess:
-    """Background process announcing every iteration of its control loop."""
-
-    def __init__(self, emitter: SignalEmitter):
-        self.emitter = emitter
-
-    def run(self, stop_reader, clock):
-        while not stop_reader.read().data:
-            self.emitter.emit('beat')
-            yield Sleep(0.01)
-
-
 class DummyControlSystem(ControlSystem):
     """Minimal control system used for integration-style tests."""
 
@@ -240,6 +228,20 @@ class TestVirtualClock:
             assert isinstance(world.clock, SystemClock)
 
 
+# Module scope: `start_in_subprocess` pickles the loop to reach a spawned child, and a class defined
+# inside the test would not pickle.
+class HeartbeatLoop:
+    """Control loop announcing every iteration of its body."""
+
+    def __init__(self, emitter: SignalEmitter):
+        self.emitter = emitter
+
+    def run(self, stop_reader, clock):
+        while not stop_reader.read().data:
+            self.emitter.emit('beat')
+            yield Sleep(0.01)
+
+
 class TestWorld:
     """Test the World class."""
 
@@ -258,7 +260,7 @@ class TestWorld:
         with world:
             emitter, receiver = world.mp_pipes()
             assert isinstance(receiver, SignalReceiver)
-            world.start_in_subprocess(HeartbeatProcess(emitter).run)
+            world.start_in_subprocess(HeartbeatLoop(emitter).run)
 
             assert len(world.background_processes) == 1
 

--- a/positronic/simulator/libero/tests/test_e2e.py
+++ b/positronic/simulator/libero/tests/test_e2e.py
@@ -13,7 +13,7 @@ The fixture is generated once on a LIBERO box::
 
 then run by explicit path (macOS renders via GLFW, no env var; Linux needs ``MUJOCO_GL=osmesa``)::
 
-    uv run --locked pytest positronic/simulator/libero/tests/test_e2e.py --no-cov
+    uv run --locked pytest positronic/simulator/libero/tests/test_e2e.py
 """
 
 import os

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -203,10 +203,6 @@ addopts = [
     # The suite waits on spawned processes and sockets about as much as it computes, so it scales with cores.
     # `-n0` puts it back in one process, which live logs and `--pdb` need.
     "-n", "auto",
-    "--cov=pimm",
-    "--cov=platform_client",
-    "--cov=positronic",
-    "--cov-report=term-missing",
     # Needs the `libero` extra and OSMesa; `libero-e2e.yaml` runs it by path, which `--ignore` allows.
     "--ignore=positronic/simulator/libero/tests/test_e2e.py",
 ]

